feat(datagrid-web): added "Load More" as configurable.

sharadsums opened the pull request:
### **Pull request type**
**New feature (non-breaking change which adds functionality)**

**Description**

Changed xml to add Load More as pagination option

rendering load more button in widget component below widget footer based on conditions that  records should not be empty and pagination selected should be Load More also added 1 condition in widget content component to stop automatic loading of records incase of LoadMore because it should be only added on click of Load More button.

Tested with filters applied and when no records are present it is working fine.

Done small styling to center align the load more button

iobuhov commented on .github/pull_request_template.md:
HI @sharadsums,
Appreciate your PR.
Just small note. Can you please copy changes in this file and add them in PR description. not in this file.
This file is just a template and it should not be changed. We can't accept your PR if this file is touched. Thank you ✋

sharadsums replied:
Hi @iobuhov thanks for letting me know, I have removed changes in template file you mentioned.

sharadsums commented:
Hi @iobuhov  any update on this? please let me know if anything is required from my side.

Thanks

gjulivan commented:
please put the text "Load More" as configurable.

sharadsums commented:
Hi @gjulivan  Have changed the title as you mentioned.
Thank you

gjulivan commented on packages/pluggableWidgets/datagrid-web/src/components/Widget.tsx:
```diff
@@ -249,6 +257,15 @@ export function Widget<C extends GridColumn>(props: WidgetProps<C>): ReactElemen
+                        onClick={() => setPage && setPage(prev => prev + 1)}
+                        tabIndex={0}
+                    >
+                        Load More
```
we usually avoid using unconfigurable text on the widget,
as the widget needs to be translate-able.
you can use textTemplate to allow user to configure it.
and show/hide this options depends on the paginationtype selection.

gjulivan commented:
> Hi @gjulivan done. Thank you

sorry, I wasn't meant the changelog, you can revert that changes,
I put the comment on the correct line of the code. please check

sharadsums commented:
Hi @gjulivanone. Thanks for your valuable feedback
 I have made load more text as configurable which will be picked from the text template as widget configuration also it is hidden when pagination type is not load more.
 Please let me know if anything i missed or required from my end.

r0b1n commented on packages/pluggableWidgets/datagrid-web/src/Datagrid.xml:
```diff
@@ -201,6 +201,7 @@
                     <enumerationValues>
                         <enumerationValue key="buttons">Paging buttons</enumerationValue>
                         <enumerationValue key="virtualScrolling">Virtual scrolling</enumerationValue>
+                        <enumerationValue key="LoadMore">Load More</enumerationValue>
```
Please use camel case notation for the value: `loadMore`.

r0b1n commented on packages/pluggableWidgets/datagrid-web/CHANGELOG.md:
```diff
@@ -6,6 +6,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
 
+
+## [2.14.0] - 2024-03-04
+
+### Added
+
+-   "Load More" as configurable.
+
```
The version shouldn't be put manually, the new section should stay under "Unreleased", the release automation will put the version automatically.

r0b1n commented on packages/pluggableWidgets/datagrid-web/src/Datagrid.xml:
```diff
@@ -220,6 +221,13 @@
                         <enumerationValue key="auto">Auto</enumerationValue>
                     </enumerationValues>
                 </property>
+                <property key="customMessage" type="textTemplate" required="false">
```
`customMessage` is a broad term, the name should reflect what this property is about and show the fact that it relates to the "Load more" pagination option.

r0b1n commented on packages/pluggableWidgets/datagrid-web/src/components/WidgetContent.tsx:
```diff
@@ -9,6 +9,7 @@ export type WidgetContentProps = {
     className?: string;
     children?: ReactNode;
     style?: React.CSSProperties;
+    paginationType?: string;
```
This props shouldn't be of type string, it should use `PaginationEnum` instead.

r0b1n commented on packages/pluggableWidgets/datagrid-web/src/Datagrid.editorPreview.tsx:
```diff
@@ -147,6 +147,7 @@ export function preview(props: DatagridPreviewProps): ReactElement {
             headerWrapperRenderer={selectableWrapperRenderer(previewColumns)}
             numberOfItems={5}
             page={0}
+            paginationType={props.pagination}
```
I am not sure how this will look in design preview but we probably should pass custom message to the component so the button it renders has proper caption.

r0b1n commented on packages/pluggableWidgets/datagrid-web/src/components/Widget.tsx:
```diff
@@ -249,6 +258,15 @@ export function Widget<C extends GridColumn>(props: WidgetProps<C>): ReactElemen
                     </Grid>
                 </WidgetContent>
                 <WidgetFooter pagination={pagination} pagingPosition={pagingPosition} />
+                {rows.length !== 0 && "LoadMore" === paginationType && (
```
Please use consistent comparison `variable === "value"`. Same in other places.

sharadsums commented:
Hi @r0b1n  I have corrected all the points you highlighted.
please have a look.
Thank you.

r0b1n commented on packages/pluggableWidgets/datagrid-web/src/components/Widget.tsx:
```diff
@@ -249,6 +258,15 @@ export function Widget<C extends GridColumn>(props: WidgetProps<C>): ReactElemen
                     </Grid>
                 </WidgetContent>
                 <WidgetFooter pagination={pagination} pagingPosition={pagingPosition} />
+                {rows.length !== 0 && paginationType === "loadMore" && (
```
I would expect the hiding logic to check if the datasource has possibility to load more items, not the number of items currently shown on the page. There is prop called `hasMoreItems` for that purpose.

r0b1n commented on packages/pluggableWidgets/datagrid-web/src/components/Widget.tsx:
```diff
@@ -249,6 +258,15 @@ export function Widget<C extends GridColumn>(props: WidgetProps<C>): ReactElemen
                 </WidgetContent>
                 <WidgetFooter pagination={pagination} pagingPosition={pagingPosition} />
+                {rows.length !== 0 && paginationType === "loadMore" && (
+                    <button
+                        className="btn btn-primary widget-datagrid-load-more"
+                        onClick={() => setPage && setPage(prev => prev + 1)}
+                        tabIndex={0}
+                    >
+                        {loadMoreMessage}
+                    </button>
```
This button should likely be located in the same place where the paging buttons are in the scenario with paging buttons (`table-footer` or something like that). So this should probably move into `WidgetFooter` component and the component has to be extended accordingly.

r0b1n commented on packages/pluggableWidgets/datagrid-web/src/components/Widget.tsx:
```diff
@@ -249,6 +258,15 @@ export function Widget<C extends GridColumn>(props: WidgetProps<C>): ReactElemen
+                        onClick={() => setPage && setPage(prev => prev + 1)}
+                        tabIndex={0}
+                    >
+                        {loadMoreMessage}
```
Strictly speaking it is not a message, it is a caption of a "load more" button. Let's name it accordingly - `loadMoreButtonCaption`, and in other places/props as well.

r0b1n commented:
Hey @sharadsums , thank you for you contribution! I resolved all the old comments. Let's iterate one more time and improve it even further. I put a few more comments/suggestions.

sharadsums commented:
Hi @r0b1n Thank you for your valuable feedbacks and insights , I have moved my button code inside the widget footer 
also i have named loadmore message to loadMoreButtonCaption , just in the textTemplate property caption i have kept "load more caption" instead of loadMoreButtonCaption as it was hiding due to length
also i have added condition hasmoreitems on top of rows.length because rows.length I am using when no records are present.

sharadsums commented:
Hi @r0b1n  good morning/afternoon
I am working on 1 other functionality for different PR and for that  i will have to modify dependency files present in node modules , is there any github repo for those dependency files.
am i missing something or what is procedure for the same.
Could you please help.
Thanks in advance.

iobuhov commented:
> Hi @r0b1n good morning/afternoon I am working on 1 other functionality for different PR and for that i will have to modify dependency files present in node modules , is there any github repo for those dependency files. am i missing something or what is procedure for the same. Could you please help. Thanks in advance.

Can you tell which dependencies you planning to change?

sharadsums commented:
Hi @iobuhov Thanks for your attention , actually in the textfilter of datagrid2 we have  options like equal to,contains,greater than ,starts with etc there i wanted to add 1 more filter type "Not contains" and these all filters types are defined in dependency files in builders and filters.
<img width="657" alt="Capture" src="https://github.com/mendix/web-widgets/assets/128319764/87040699-dea9-4ad4-9bc2-07fa08f4bdcd">

sharadsums commented:
Hi @r0b1n @leonardomendix  thanks for approving , justmodified  change log file for conflicts in this commit
